TS2322: Type 'Input' is not assignable to type 'FormValueControl<unknown>'

### Which @angular/* package(s) are the source of the bug?

forms

### Is this a regression?

Yes

### Description

The recently added template type checking for signal forms #64823 prevents FormValueControl to be bound to field directive.

Minimal example:

```ts
@Component({ selector: 'app-input', template: '' })
class Input implements FormValueControl<string> {
  value = model('');
}

@Component({
  selector: 'app-root',
  imports: [Field, Input],
  template: `<app-input [field]="field" />`,
})
class App {
  field = form(signal(''));
}
``` 

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/stackblitz-starters-9hbcnbex?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
✘ [ERROR] TS2322: Type 'Input' is not assignable to type 'FormValueControl<unknown>'.
  The types of 'value[SIGNAL].transformFn' are incompatible between these types.
    Type '((value: string) => string) | undefined' is not assignable to type '((value: unknown) => unknown) | undefined'.
      Type '(value: string) => string' is not assignable to type '(value: unknown) => unknown'.
        Types of parameters 'value' and 'value' are incompatible.
          Type 'unknown' is not assignable to type 'string'. [plugin angular-compiler]

    src/main.ts:18:24:
      18 │   template: `<app-input [field]="field" />`,
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI       : 21.0.0-rc.1
Angular           : 21.0.0-rc.1
Node.js           : 20.19.1
Package Manager   : npm 10.8.2
Operating System  : linux x64

┌───────────────────────────┬───────────────────┬───────────────────┐
│ Package                   │ Installed Version │ Requested Version │
├───────────────────────────┼───────────────────┼───────────────────┤
│ @angular/build            │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/cli              │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/common           │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/compiler         │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/compiler-cli     │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/core             │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/forms            │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/platform-browser │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ @angular/router           │ 21.0.0-rc.1       │ 21.0.0-rc.1       │
│ rxjs                      │ 7.8.2             │ ^7.8.1            │
│ typescript                │ 5.9.2             │ ^5.9.0            │
└───────────────────────────┴───────────────────┴───────────────────┘
```

### Anything else?

Even casting the field value as `any` (or `unknown`):`field = form(signal<any>(''))`  or using `any` as a generic for `FormValueControl` does not provide a workaround.

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
@@ -746,7 +746,28 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
  */
 abstract class TcbFieldDirectiveTypeBaseOp extends TcbOp {
   /** Bindings that aren't supported on signal form fields. */
-  private unsupportedBindingFields: Set<string>;
+  protected readonly unsupportedBindingFields = new Set([
+    // Should be kept in sync with the `FormUiControl` bindings,
+    // defined in `packages/forms/signals/src/api/control.ts`.
+    'value',
+    'checked',
+    'errors',
+    'invalid',
+    'disabled',
+    'disabledReasons',
+    'name',
+    'readonly',
+    'touched',
+    'max',
+    'maxlength',
+    'maxLength',
+    'min',
+    'minLength',
+    'minlength',
+    'pattern',
+    'required',
+    'type',
+  ]);
 
   constructor(
     protected tcb: Context,
@@ -755,32 +776,6 @@ abstract class TcbFieldDirectiveTypeBaseOp extends TcbOp {
     protected dir: TypeCheckableDirectiveMeta,
   ) {
     super();
-
-    // Should be kept in sync with the `FormUiControl` bindings,
-    // defined in `packages/forms/signals/src/api/control.ts`.
-    const commonUnsupportedNames = [
-      'value',
-      'checked',
-      'errors',
-      'invalid',
-      'disabled',
-      'disabledReasons',
-      'name',
-      'readonly',
-      'touched',
-      'max',
-      'maxlength',
-      'maxLength',
-      'min',
-      'minLength',
-      'minlength',
-      'pattern',
-      'required',
-      'type',
-    ];
-
-    // `type` can't be bound, but is allowed as a static attribute.
-    this.unsupportedBindingFields = new Set(commonUnsupportedNames);
   }
 
   override get optional() {
@@ -840,6 +835,23 @@ abstract class TcbFieldDirectiveTypeBaseOp extends TcbOp {
  * A `TcbOp` which constructs an instance of the signal forms `Field` directive on a native element.
  */
 class TcbNativeFieldDirectiveTypeOp extends TcbFieldDirectiveTypeBaseOp {
+  private readonly inputType: string | null;
+
+  constructor(tcb: Context, scope: Scope, node: DirectiveOwner, dir: TypeCheckableDirectiveMeta) {
+    super(tcb, scope, node, dir);
+
+    this.inputType =
+      (node instanceof TmplAstElement &&
+        node.name === 'input' &&
+        node.attributes.find((attr) => attr.name === 'type')?.value) ||
+      null;
+
+    // Radio control are allowed to set the `value`.
+    if (this.inputType === 'radio') {
+      this.unsupportedBindingFields.delete('value');
+    }
+  }
+
   protected override getExpectedType(): ts.TypeNode {
     if (this.node instanceof TmplAstElement) {
       return this.getExpectedTypeFromDomNode(this.node);
@@ -858,9 +870,7 @@ class TcbNativeFieldDirectiveTypeOp extends TcbFieldDirectiveTypeBaseOp {
       return this.getUnsupportedType();
     }
 
-    const inputType = node.attributes.find((attr) => attr.name === 'type')?.value;
-
-    switch (inputType) {
+    switch (this.inputType) {
       case 'checkbox':
         return ts.factory.createKeywordTypeNode(ts.SyntaxKind.BooleanKeyword);
 

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
@@ -975,7 +975,7 @@ class TcbCustomFieldDirectiveTypeOp extends TcbFieldDirectiveTypeBaseOp {
     const id = this.tcb.allocateId();
     const targetType = ts.factory.createTypeReferenceNode(
       targetTypeRef.typeName,
-      isCheckbox ? undefined : [ts.factory.createKeywordTypeNode(ts.SyntaxKind.UnknownKeyword)],
+      isCheckbox ? undefined : [ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)],
     );
     this.scope.addStatement(tsDeclareVariable(id, targetType));
 

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
@@ -1205,6 +1205,7 @@ class TcbDirectiveInputsOp extends TcbOp {
     private scope: Scope,
     private node: TmplAstTemplate | TmplAstElement | TmplAstComponent | TmplAstDirective,
     private dir: TypeCheckableDirectiveMeta,
+    private ignoredRequiredInputs: Set<string> | null,
   ) {
     super();
   }
@@ -1375,7 +1376,12 @@ class TcbDirectiveInputsOp extends TcbOp {
     const missing: BindingPropertyName[] = [];
 
     for (const input of this.dir.inputs) {
-      if (input.required && !seenRequiredInputs.has(input.classPropertyName)) {
+      if (
+        input.required &&
+        !seenRequiredInputs.has(input.classPropertyName) &&
+        (this.ignoredRequiredInputs === null ||
+          !this.ignoredRequiredInputs.has(input.bindingPropertyName))
+      ) {
         missing.push(input.bindingPropertyName);
       }
     }
@@ -3120,7 +3126,22 @@ class Scope {
     const directiveOp = this.getDirectiveOp(dir, node, allDirectiveMatches);
     const dirIndex = this.opQueue.push(directiveOp) - 1;
     dirMap.set(dir, dirIndex);
-    this.opQueue.push(new TcbDirectiveInputsOp(this.tcb, this, node, dir));
+
+    let ignoredRequiredInputs: Set<string> | null = null;
+
+    // The `Field` directive will bind implicitly to
+    // the relevant input so we don't need to check for it.
+    if (allDirectiveMatches.some(isFieldDirective)) {
+      const customFieldType = getCustomFieldDirectiveType(dir);
+
+      if (customFieldType === 'value') {
+        ignoredRequiredInputs = new Set(['value']);
+      } else if (customFieldType === 'checkbox') {
+        ignoredRequiredInputs = new Set(['checked']);
+      }
+    }
+
+    this.opQueue.push(new TcbDirectiveInputsOp(this.tcb, this, node, dir, ignoredRequiredInputs));
   }
 
   private getDirectiveOp(
@@ -3130,10 +3151,7 @@ class Scope {
   ): TcbOp {
     const dirRef = dir.ref as Reference<ClassDeclaration<ts.ClassDeclaration>>;
 
-    if (
-      dir.name === 'Field' &&
-      dirRef.bestGuessOwningModule?.specifier === '@angular/forms/signals'
-    ) {
+    if (isFieldDirective(dir)) {
       let customControl: {type: 'value' | 'checkbox'; meta: TypeCheckableDirectiveMeta} | null =
         null;
 
@@ -4066,22 +4084,25 @@ function getComponentTagName(node: TmplAstComponent): string {
   return node.tagName || 'ng-component';
 }
 
-/** Determines the type of signal form field control (if any) from a directive's metadata. */
-function getCustomFieldDirectiveType({
-  inputs,
-  outputs,
-}: TypeCheckableDirectiveMeta): 'value' | 'checkbox' | null {
-  if (
-    inputs.getByBindingPropertyName('value')?.some((v) => v.isSignal) &&
-    outputs.hasBindingPropertyName('valueChange')
-  ) {
-    return 'value';
-  }
+function isFieldDirective(meta: TypeCheckableDirectiveMeta): boolean {
+  return (
+    meta.name === 'Field' && meta.ref.bestGuessOwningModule?.specifier === '@angular/forms/signals'
+  );
+}
 
-  if (
-    inputs.getByBindingPropertyName('checked')?.some((v) => v.isSignal) &&
-    outputs.hasBindingPropertyName('checkedChange')
-  ) {
+function hasModel(name: string, meta: TypeCheckableDirectiveMeta): boolean {
+  return (
+    !!meta.inputs.getByBindingPropertyName(name)?.some((v) => v.isSignal) &&
+    meta.outputs.hasBindingPropertyName(name + 'Change')
+  );
+}
+
+function getCustomFieldDirectiveType(
+  meta: TypeCheckableDirectiveMeta,
+): 'value' | 'checkbox' | null {
+  if (hasModel('value', meta)) {
+    return 'value';
+  } else if (hasModel('checked', meta)) {
     return 'checkbox';
   }
 

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
@@ -747,26 +747,12 @@ class TcbGenericDirectiveTypeWithAnyParamsOp extends TcbDirectiveTypeOpBase {
 abstract class TcbFieldDirectiveTypeBaseOp extends TcbOp {
   /** Bindings that aren't supported on signal form fields. */
   protected readonly unsupportedBindingFields = new Set([
-    // Should be kept in sync with the `FormUiControl` bindings,
-    // defined in `packages/forms/signals/src/api/control.ts`.
+    ...formControlInputFields,
     'value',
     'checked',
-    'errors',
-    'invalid',
-    'disabled',
-    'disabledReasons',
-    'name',
-    'readonly',
-    'touched',
-    'max',
-    'maxlength',
-    'maxLength',
-    'min',
-    'minLength',
-    'minlength',
-    'pattern',
-    'required',
     'type',
+    'maxlength',
+    'minlength',
   ]);
 
   constructor(
@@ -3144,10 +3130,14 @@ class Scope {
     if (allDirectiveMatches.some(isFieldDirective)) {
       const customFieldType = getCustomFieldDirectiveType(dir);
 
-      if (customFieldType === 'value') {
-        ignoredRequiredInputs = new Set(['value']);
-      } else if (customFieldType === 'checkbox') {
-        ignoredRequiredInputs = new Set(['checked']);
+      if (customFieldType !== null) {
+        ignoredRequiredInputs = new Set(formControlInputFields);
+
+        if (customFieldType === 'value') {
+          ignoredRequiredInputs.add('value');
+        } else if (customFieldType === 'checkbox') {
+          ignoredRequiredInputs.add('checked');
+        }
       }
     }
 
@@ -4123,6 +4113,25 @@ function hasModel(name: string, meta: TypeCheckableDirectiveMeta): boolean {
     meta.outputs.hasBindingPropertyName(name + 'Change')
   );
 }
+
+/** Names of the input fields on custom controls. */
+const formControlInputFields = [
+  // Should be kept in sync with the `FormUiControl` bindings,
+  // defined in `packages/forms/signals/src/api/control.ts`.
+  'errors',
+  'invalid',
+  'disabled',
+  'disabledReasons',
+  'name',
+  'readonly',
+  'touched',
+  'max',
+  'maxLength',
+  'min',
+  'minLength',
+  'pattern',
+  'required',
+] as const;
 
 function getCustomFieldDirectiveType(
   meta: TypeCheckableDirectiveMeta,

--- a/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts
@@ -4095,8 +4095,25 @@ function getComponentTagName(node: TmplAstComponent): string {
 }
 
 function isFieldDirective(meta: TypeCheckableDirectiveMeta): boolean {
+  if (meta.name !== 'Field') {
+    return false;
+  }
+
+  // Fast path, relevant for all external users.
+  if (meta.ref.bestGuessOwningModule?.specifier === '@angular/forms/signals') {
+    return true;
+  }
+
+  // Slightly slower, but more accurate path.
   return (
-    meta.name === 'Field' && meta.ref.bestGuessOwningModule?.specifier === '@angular/forms/signals'
+    ts.isClassDeclaration(meta.ref.node) &&
+    meta.ref.node.members.some(
+      (member) =>
+        ts.isPropertyDeclaration(member) &&
+        ts.isComputedPropertyName(member.name) &&
+        ts.isIdentifier(member.name.expression) &&
+        member.name.expression.text === 'ɵCONTROL',
+    )
   );
 }
 

--- a/packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts
+++ b/packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts
@@ -2658,7 +2658,7 @@ describe('type check blocks', () => {
       expect(block).toContain(
         'var _t1 = null! as i0.Field<i1.ɵExtractFormControlValue<i0.CustomControl>>;',
       );
-      expect(block).toContain('var _t2 = null! as i2.FormValueControl<unknown>;');
+      expect(block).toContain('var _t2 = null! as i2.FormValueControl<any>;');
       expect(block).toContain('if (_t2) _t2 = null! as i0.CustomControl;');
       expect(block).toContain('_t1.field = (((this).f));');
     });

--- a/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
+++ b/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
@@ -271,6 +271,33 @@ runInEachFileSystem(() => {
       );
     });
 
+    it('should allow binding to `value` on radio controls', () => {
+      env.write(
+        'test.ts',
+        `
+          import {Component, signal} from '@angular/core';
+          import {Field, form} from '@angular/forms/signals';
+
+          @Component({
+            template: \`
+              <form>
+                <input type="radio" value="a" [field]="f">
+                <input type="radio" value="b" [field]="f">
+                <input type="radio" value="c" [field]="f">
+              </form>
+            \`,
+            imports: [Field]
+          })
+          export class Comp {
+            f = form(signal('a'), {name: 'test'});
+          }
+        `,
+      );
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(0);
+    });
+
     it('should report unsupported static attributes of a field', () => {
       env.write(
         'test.ts',

--- a/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
+++ b/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
@@ -360,5 +360,57 @@ runInEachFileSystem(() => {
         `The types of 'required[SIGNAL].transformFn' are incompatible between these types.`,
       );
     });
+
+    it('should not report `value` as a missing required input when the `Field` directive is present', () => {
+      env.write(
+        'test.ts',
+        `
+          import {Component, signal, model} from '@angular/core';
+          import {Field, form, FormValueControl} from '@angular/forms/signals';
+
+          @Component({selector: 'custom-control', template: ''})
+          export class CustomControl implements FormValueControl<string> {
+            readonly value = model.required<string>();
+          }
+
+          @Component({
+            template: '<custom-control [field]="f"/>',
+            imports: [Field, CustomControl]
+          })
+          export class Comp {
+            f = form(signal(''));
+          }
+        `,
+      );
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(0);
+    });
+
+    it('should not report `checked` as a missing required input when the `Field` directive is present', () => {
+      env.write(
+        'test.ts',
+        `
+          import {Component, signal, model} from '@angular/core';
+          import {Field, form, FormCheckboxControl} from '@angular/forms/signals';
+
+          @Component({selector: 'custom-control', template: ''})
+          export class CustomControl implements FormCheckboxControl {
+            readonly checked = model.required<boolean>();
+          }
+
+          @Component({
+            template: '<custom-control [field]="f"/>',
+            imports: [Field, CustomControl]
+          })
+          export class Comp {
+            f = form(signal(false));
+          }
+        `,
+      );
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(0);
+    });
   });
 });

--- a/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
+++ b/packages/compiler-cli/test/ngtsc/signal_forms_spec.ts
@@ -19,7 +19,11 @@ runInEachFileSystem(() => {
 
     beforeEach(() => {
       env = NgtscTestEnvironment.setup(testFiles);
-      env.tsconfig({strictTemplates: true});
+      env.tsconfig({
+        // This is the default in Angular apps so we enable it to ensure consistent behavior.
+        strict: true,
+        strictTemplates: true,
+      });
     });
 
     function extractMessage(diag: ts.Diagnostic) {
@@ -191,6 +195,33 @@ runInEachFileSystem(() => {
       );
     });
 
+    it('should not report a custom control that conforms to `FormValueControl`', () => {
+      env.write(
+        'test.ts',
+        `
+          import {Component, signal, model} from '@angular/core';
+          import {Field, form, FormValueControl} from '@angular/forms/signals';
+
+          @Component({ selector: 'string-control', template: '' })
+          class StringControl implements FormValueControl<string> {
+            value = model('');
+          }
+
+          @Component({
+            selector: 'app-root',
+            imports: [Field, StringControl],
+            template: '<string-control [field]="field" />',
+          })
+          class App {
+            field = form(signal(''));
+          }
+        `,
+      );
+
+      const diags = env.driveDiagnostics();
+      expect(diags.length).toBe(0);
+    });
+
     it('should report unsupported property bindings on a field', () => {
       env.write(
         'test.ts',
@@ -290,10 +321,10 @@ runInEachFileSystem(() => {
       const diags = env.driveDiagnostics();
       expect(diags.length).toBe(1);
       expect(extractMessage(diags[0])).toBe(
-        `Type 'UserControl' is not assignable to type 'FormValueControl<unknown>'.`,
+        `Type 'UserControl' is not assignable to type 'FormValueControl<any>'.`,
       );
       expect((diags[0].messageText as ts.DiagnosticMessageChain).next?.[0].messageText).toBe(
-        `Types of property 'required' are incompatible.`,
+        `The types of 'required[SIGNAL].transformFn' are incompatible between these types.`,
       );
     });
 
@@ -326,7 +357,7 @@ runInEachFileSystem(() => {
         `Type 'UserControl' is not assignable to type 'FormCheckboxControl'.`,
       );
       expect((diags[0].messageText as ts.DiagnosticMessageChain).next?.[0].messageText).toBe(
-        `Types of property 'required' are incompatible.`,
+        `The types of 'required[SIGNAL].transformFn' are incompatible between these types.`,
       );
     });
   });

--- a/packages/forms/signals/test/web/field_directive.spec.ts
+++ b/packages/forms/signals/test/web/field_directive.spec.ts
@@ -115,7 +115,7 @@ describe('field directive', () => {
         })
         class TestCmp {
           readonly disabled = signal(false);
-          readonly f = form(signal(false), (p) => {
+          readonly f = form(signal(''), (p) => {
             disabled(p, this.disabled);
           });
         }
@@ -376,26 +376,6 @@ describe('field directive', () => {
         expect(element.required).toBe(true);
       });
 
-      it('should bind to native control and override attribute', () => {
-        @Component({
-          imports: [Field],
-          template: `<input [field]="f" required>`,
-        })
-        class TestCmp {
-          readonly required = signal(false);
-          readonly f = form(signal(''), (p) => {
-            required(p, {when: this.required});
-          });
-        }
-
-        const fixture = act(() => TestBed.createComponent(TestCmp));
-        const element = fixture.nativeElement.firstChild;
-        expect(element.required).withContext("'required' should be overridden").toBe(false);
-
-        act(() => fixture.componentInstance.required.set(true));
-        expect(element.required).toBe(true);
-      });
-
       it('should bind to custom control', () => {
         @Component({selector: 'custom-control', template: ``})
         class CustomControl implements FormValueControl<string> {
@@ -520,22 +500,6 @@ describe('field directive', () => {
         expect(component.customControl().max()).toBe(5);
       });
 
-      it('should not bind to native control that does not support it', () => {
-        @Component({
-          imports: [Field],
-          template: `<input type="text" [field]="f">`,
-        })
-        class TestCmp {
-          readonly f = form(signal(5), (p) => {
-            max(p, 10);
-          });
-        }
-
-        const fixture = act(() => TestBed.createComponent(TestCmp));
-        const element = fixture.nativeElement.firstChild as HTMLInputElement;
-        expect(element.max).toBe('');
-      });
-
       it('should be reset when field changes on native control', () => {
         @Component({
           imports: [Field],
@@ -631,22 +595,6 @@ describe('field directive', () => {
 
         act(() => component.min.set(5));
         expect(component.customControl().min()).toBe(5);
-      });
-
-      it('should not bind to native control that does not support it', () => {
-        @Component({
-          imports: [Field],
-          template: `<input type="text" [field]="f">`,
-        })
-        class TestCmp {
-          readonly f = form(signal(15), (p) => {
-            min(p, 10);
-          });
-        }
-
-        const fixture = act(() => TestBed.createComponent(TestCmp));
-        const element = fixture.nativeElement.firstChild as HTMLInputElement;
-        expect(element.min).toBe('');
       });
 
       it('should be reset when field changes on native control', () => {
@@ -1268,7 +1216,7 @@ describe('field directive', () => {
 
     @Component({
       imports: [Field, CustomInput],
-      template: `<my-input [field]="f" value />`,
+      template: `<my-input [field]="f" />`,
     })
     class TestCmp {
       f = form<string>(signal('test'));

